### DISCUSSION TITLE
系统中依赖库升级后之前编译过的项目不能正常编译

### DISCUSSION THREAD

**Discussion by huye** [xmake-io/xmake#2199] [Ideas]
### Xmake 版本

v2.6.4+dev.35fa885

### 操作系统版本和架构

macOS 12.2.1  2015 amd64

### 描述问题

今天更新了下 brew，结果昨天能编译的项目又不能编译了，看了下是xmake缓存了之前的旧版依赖库的路径。现在我直接 rm -rf .xmake 删除缓存来解决的，但是这样的话，就要全部重新编译，如果是大项目比较浪费时间。

### 期待的结果

希望能自动判断更新外部依赖版本，重新链接就好了，不用全部重新编译。

**Comment by xq114** [comment]
xmake f -c可以清除缓存；更新依赖之后本来就要重新编译的，依赖库变了不可能还和原来的文件链接到一起

**Comment by huye** [comment]
这样吗，我直接 xmake clean 清不掉。只是依赖库更新，接口没有变应该是可以不重新编译的吧，只需要重新链接就好了。

**Comment by xq114** [comment]
> 这样吗，我直接 xmake clean 清不掉。只是依赖库更新，接口没有变应该是可以不重新编译的吧，只需要重新链接就好了。

xmake clean只是清除中间文件；xmake config -c才是更新依赖。接口变没变这个xmake无法确定的，更新依赖不会太频繁，偶尔重新编译一下问题不大

**Comment by huye** [comment]
接口变了的话，肯定要修改源文件啊，这个不就能判断了。make应该就是基于文件修改时间来判断的。如果接口变了又没有重新编译，这个链接的时候也会报错，这个时候就算忘了修改源码，应该也能想起来了。

**Comment by xq114** [comment]
> 接口变了的话，肯定要修改源文件啊，这个不就能判断了。make应该就是基于文件修改时间来判断的。如果接口变了又没有重新编译，这个链接的时候也会报错，这个时候就算忘了修改源码，应该也能想起来了。

不是这么回事，你说的是API接口，我说的是ABI接口，很多库更新的时候API接口不变但ABI接口会变的，没法检测，比如你用了public的funcA函数，本来头文件里funcA调用funcB，现在funcA调用funcC，你不重新编译直接链接就会报错找不到funcB

**Comment by waruqi** [comment]
更新 xmake 会导致各种可能性的变动，包括缓存格式，中间数据的存储路径变动 (尽管不可能每次都变)，我记得更新完 xmake 会强制清一次 config cache ，重新检测所有，这个是躲不掉的。。如果自动清没生效，也建议自己手动执行 xmake f -c 强制触发检测所有缓存信息。

否则可能会导致各种问题。。而且又不可能天天升级 xmake。。

**Reply by waruqi** [reply]
> 这样吗，我直接 xmake clean 清不掉。只是依赖库更新，接口没有变应该是可以不重新编译的吧，只需要重新链接就好了。
> 不是这么回事，你说的是API接口，我说的是ABI接口，很多库更新的时候API接口不变但ABI接口会变的，没法检测，比如你用了public的funcA函数，本来头文件里funcA调用funcB，现在funcA调用funcC，你不重新编译直接链接就会报错找不到funcB

跟接口没关系，每个依赖库都有 buildhash 唯一标识依赖 configs， 版本，平台架构，工具链等信息，这样依赖库的版本和所有其他信息都不变，即使你删了 .xmake 或者执行 `xmake f -c` 也仅仅导致重新检测依赖，只要 buildhash 和 version 不变，是不会重编的。。

按你现在的描述出现重编的情况：

1. 升级 xmake，会强制触发一次 xmake cache 清除，或者你删了 .xmake 都会触发重新检测所有依赖
2. 这个时候 检测到你依赖的库，有版本更新，会使用新的版本，因为 add_requires("xxx") 默认是 latest 版本选择，不是锁死版本的
3. 检测到新版本，就会重新下载编译依赖，因为你的系统上只安装了老版本依赖包

所以跟 abi 没任何关系，你想不重编译依赖，就锁定依赖版本 ，可以指定每个依赖版本，不要使用默认的 latest。。或者使用 requires_lock 
 依赖版本锁定策略。https://github.com/xmake-io/xmake/issues/1567

**Reply by huye** [reply]
我感觉老大好像理解反了，或者是我之前可能两个问题说混了。我之前遇到的情况是昨天编译的时候系统里安装的是 mariadb-connector-c@3.2.5，今天更新了下成了 mariadb-connector-c@3.2.6，不知道为什么，我重新编译他没有更新依赖，不记得是编译还是链接的时候报错，提示找不到 mariadb-connector-c/3.2.5 的相关文件。

**Reply by huye** [reply]
不知道这次我说清楚没有，总之就是我昨天能正常编译，今天更新了库后，再编译报错，xmake clean 后再编译还是报错，我手动 rm -f .xmake 后再编译正常，就是这么回事。

**Reply by waruqi** [reply]
xmake-repo 仓库这个包只有 3.1.13 。。没有 3.2.5/3.2.6 https://github.com/xmake-io/xmake-repo/blob/5fc332057f5915f9fd1dde82c41cd09c14b10970/packages/m/mariadb-connector-c/xmake.lua#L6

如果你用的 conan ，conan 包的依赖信息是缓存到工程本地的，我不知道你是怎么更新 conan 包的，如果走 conan 命令更新的，肯定要 xmake f -c 清下重新检测的，否则用的肯定还是老的 3.2.5，如果是 xmake.lua 里改成了 3.2.6 。。会自动拉新版本编译集成

**Reply by huye** [reply]
我还没有用仓库管理包，我是用 brew 安装的，所以之前才有说用到 pkg-config 这些的，现在改为了使用 add_requires。现在杂事太多了，这里弄一下，哪里弄一下，过一会都忘了刚才再做什么了。还是习惯使用以前就会的方法管理，现在是能不新学尽量不新学。特别是有时候新学的东西和以前用过的功能一些有交叉，更容易混乱。

**Comment by huye** [comment]
上面说的那个应该是说像那种宏是一样的，但是宏的值改变了这样吧，这个确实是我没想到。那个 xmake f -c 我是没有想到这样去清除配置缓存，我以为xmake clean就能清除了。那是不直接再增加个类似 xmake clean -f 或 -a 的（force, all 这样），清除功能在一起。另外像 git 一样，执行结果给一些相关功能的提示，有时可能没想到用哪个功能，提示一下就知道了。git 真的是用过的最好用的功能。基本功能都不用去看手册。

**Reply by waruqi** [reply]
xmake config 处理配置，xmake clean 处理构建产物，目前不会做调整。。文档有说明

**Reply by waruqi** [reply]
另外，`xmake clean -a` 可以删掉所有，.xmake 等一切自动生成的文件。。这原本就是存在的特性，要学会自己先翻 --help，包括 --help 也有说明

**Reply by huye** [reply]
好的好的，哎现在项目多了，换来换去的弄，之前看过的后来也会忘掉。我本来不是反馈添加这个功能的，哈哈，就当我没提过这个 xmake clean -a。
